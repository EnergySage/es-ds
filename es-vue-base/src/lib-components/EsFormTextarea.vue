<template>
    <div
        :required="required">
        <!-- eslint-disable-next-line vuejs-accessibility/label-has-for -->
        <label
            :for="id"
            class="label font-weight-bold">
            <slot name="label" />

            <span
                v-if="required"
                class="text-danger">
                *
            </span>
        </label>
        <div>
            <b-form-textarea
                :id="id"
                class="es-form-textarea"
                :disabled="disabled"
                :state="state" />
            <b-form-text v-if="state == null">
                <slot name="message" />
            </b-form-text>
            <b-form-invalid-feedback>
                <slot name="errorMessage" />
            </b-form-invalid-feedback>
            <b-form-valid-feedback>
                <slot name="successMessage" />
            </b-form-valid-feedback>
        </div>
    </div>
</template>

<script lang="js">
import {
    BFormText,
    BFormTextarea,
    BFormInvalidFeedback,
    BFormValidFeedback,
} from 'bootstrap-vue';

export default {
    name: 'EsFormTextarea',
    components: {
        BFormText,
        BFormTextarea,
        BFormInvalidFeedback,
        BFormValidFeedback,
    },
    props: {
        /**
         * Required
         */
        required: {
            type: Boolean,
            default: false,
        },
        /**
         * Disabled
         */
        disabled: {
            type: Boolean,
            default: false,
        },
        /**
         * ID
         * required
         */
        id: {
            type: String,
            required: true,
        },
        /**
         * state
         */
        state: {
            type: [Boolean, null],
            default: null,
        },
    },
};
</script>

<style lang="scss" scoped>
@import '~@energysage/es-bs-extends/scss/includes';

.es-form-textarea {
    min-height: 8.125rem;
}
</style>
