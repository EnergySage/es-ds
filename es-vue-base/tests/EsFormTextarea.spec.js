import { mount } from '@vue/test-utils';
import EsFormTextarea from '@/src/lib-components/molecules/EsFormTextarea.vue';
import jestVue from '@/tests/jest.vue.config';

describe('EsFormTextarea', () => {
    // Basic test; should exist for most components
    test('<EsFormTextarea />', async () => {
        const wrapper = mount(EsFormTextarea, {
            ...jestVue,
            slots: {
                label: 'Notes',
            },
            propsData: {
                id: 'myID',
            },
        });
        const a11y = await axe(wrapper.element);

        expect(wrapper.vm).toBeTruthy();
        expect(wrapper.html()).toMatchSnapshot();
        expect(a11y).toHaveNoViolations();
    });

    // Test slots are there
    test('<EsFormTextarea>slots</EsFormTextarea>', async () => {
        const wrapper = mount(EsFormTextarea, {
            ...jestVue,
            slots: {
                label: 'Notes',
                message: 'Enter your notes here.',
                errorMessage: 'Please enter a valid value.',
                successMessage: 'Your response has been submitted successfully.',
            },
            propsData: {
                id: 'myID',

            },
        });
        expect(wrapper.vm).toBeTruthy();
        expect(wrapper.html()).toMatchSnapshot();
        expect(wrapper.vm.$slots.label[0].text).toBe('Notes');
        expect(wrapper.vm.$slots.message[0].text).toBe('Enter your notes here.');
        expect(wrapper.vm.$slots.errorMessage[0].text).toBe('Please enter a valid value.');
        expect(wrapper.vm.$slots.successMessage[0].text).toBe('Your response has been submitted successfully.');
    });

    // Test label structure
    test('<EsFormTextarea <label /> />', async () => {
        const wrapper = mount(EsFormTextarea, {
            ...jestVue,
            propsData: {
                id: 'myID',
                required: true,
            },
        });

        expect(wrapper.vm).toBeTruthy();
        expect(wrapper.html()).toMatchSnapshot();

        const label = wrapper.find('label');
        expect(label.exists()).toBe(true);
        expect(label.find('*').exists()).toBe(true);
    });

    // Test props
    test('<EsFormTextarea props />', async () => {
        const wrapper = mount(EsFormTextarea, {
            ...jestVue,
            propsData: {
                id: 'myID',
                required: true,
            },
        });

        expect(wrapper.vm).toBeTruthy();
        expect(wrapper.html()).toMatchSnapshot();

        expect(wrapper.props('id')).toBe('myID');
        expect(wrapper.props('state')).toBeNull();
    });
});
