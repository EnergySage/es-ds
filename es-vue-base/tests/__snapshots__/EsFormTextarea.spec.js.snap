// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EsFormTextarea <EsFormTextarea /> 1`] = `
"<div class="input-wrapper justify-content-end mb-2"><label for="myID" class="label font-weight-semibold justify-content-start">Notes
    <!----></label>
  <div class="input-holder"><textarea id="myID" rows="2" wrap="soft" class="es-form-textarea w-100 form-control"></textarea>
    <!---->
    <!---->
    <!---->
  </div>
</div>"
`;

exports[`EsFormTextarea <EsFormTextarea <label /> /> 1`] = `
"<div required="required" class="input-wrapper justify-content-end mb-2"><label for="myID" class="label font-weight-semibold justify-content-start"> <span class="text-danger">
            *
        </span></label>
  <div class="input-holder"><textarea id="myID" rows="2" wrap="soft" class="es-form-textarea w-100 form-control"></textarea>
    <!---->
    <div class="invalid-feedback">
      This field is required.
    </div>
    <!---->
  </div>
</div>"
`;

exports[`EsFormTextarea <EsFormTextarea props /> 1`] = `
"<div required="required" class="input-wrapper justify-content-end mb-2"><label for="myID" class="label font-weight-semibold justify-content-start"> <span class="text-danger">
            *
        </span></label>
  <div class="input-holder"><textarea id="myID" rows="2" wrap="soft" class="es-form-textarea w-100 form-control"></textarea>
    <!---->
    <div class="invalid-feedback">
      This field is required.
    </div>
    <!---->
  </div>
</div>"
`;

exports[`EsFormTextarea <EsFormTextarea v-model /> 1`] = `
"<div>
  <div class="input-wrapper justify-content-end mb-2"><label for="testId" class="label font-weight-semibold justify-content-start">
      <!----></label>
    <div class="input-holder"><textarea id="testId" rows="2" wrap="soft" class="es-form-textarea w-100 form-control" data-testid="testTextarea"></textarea>
      <!---->
      <!---->
      <!---->
    </div>
  </div>
</div>"
`;

exports[`EsFormTextarea <EsFormTextarea>slots</EsFormTextarea> 1`] = `
"<div class="input-wrapper justify-content-end mb-2"><label for="myID" class="label font-weight-semibold justify-content-start">Notes
    <!----></label>
  <div class="input-holder"><textarea id="myID" rows="2" wrap="soft" class="es-form-textarea w-100 form-control"></textarea>
    <!---->
    <div class="invalid-feedback">Please enter a valid value.</div>
    <div class="valid-feedback">Your response has been submitted successfully.</div>
  </div>
</div>"
`;
